### PR TITLE
feat: add discv4 terminate

### PR DESCRIPTION
**Pull request by DoTheBestToGetTheBest**
Draft for https://github.com/paradigmxyz/reth/issues/4874

### CODE REVIEW

**Comment by DoTheBestToGetTheBest** [comment]
Could i ask help on what i need to do in the next step please?

File: crates/net/discv4/src/lib.rs [hunk 1]
```diff
@@ -711,8 +716,8 @@ impl Discv4Service {
             self.kbuckets
                 .closest_values(&target_key)
                 .filter(|node| {
-                    node.value.has_endpoint_proof &&
-                        !self.pending_find_nodes.contains_key(&node.key.preimage().0)
+                    node.value.has_endpoint_proof
+                        && !self.pending_find_nodes.contains_key(&node.key.preimage().0)
```
Comment (mattsse) [inline crates/net/discv4/src/lib.rs:720]: please use nightly formatting `cargo +nightly fmt`

File: crates/net/discv4/src/lib.rs [hunk 2]
```diff
@@ -1554,6 +1559,11 @@ impl Discv4Service {
                             let _ = self.local_eip_868_enr.set_tcp6(port, &self.secret_key);
                         }
                     }
+
+                    Discv4Command::Terminated => {
+                        //self.terminate();
+                        todo!()
```
Comment (mattsse) [inline crates/net/discv4/src/lib.rs:1565]: this needs to emit the event 

and here we need to handle the terminate:

https://github.com/paradigmxyz/reth/blob/858ea41da1f8852e93e0604dbb2f7a89a5398b28/crates/net/discv4/src/lib.rs#L1618-L1618

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by DoTheBestToGetTheBest** [comment]
I'm really sorry if i'm asking newbie questions for most of you

**Comment by mattsse** [comment]
it's okay, we're almost there, the only thing missing:

https://github.com/paradigmxyz/reth/pull/4879#discussion_r1342963406

**Comment by DoTheBestToGetTheBest** [comment]
> it's okay, we're almost there, the only thing missing:
> 
> [#4879 (comment)](https://github.com/paradigmxyz/reth/pull/4879#discussion_r1342963406)

I'm really sorry but i didn't what it means by emit the event ? Sorry again Mat, i know you have better to do then telling me

**Comment by DoTheBestToGetTheBest** [comment]
@mattsse hey again, sorry again for all this questions. How it look like for you ? added an event, and return None if Poll::ready is terminated

**Review by mattsse — APPROVED** [review]
this is great, tysm!

**Comment by DoTheBestToGetTheBest** [comment]
> this is great, tysm!

oh i don't understand this, is this okey ? or i need to change others thing or i'm wrong in the implementation ?

**Comment by mattsse** [comment]
yes this is great, I made a few smol touchups and pushed to #4888

because you're opened the PR from your main branch I'm unable to push changes here directly,
if you want you can merge #4888 into this PR real quick

**Comment by DoTheBestToGetTheBest** [comment]
> yes this is great, I made a few smol touchups and pushed to #4888
> 
> because you're opened the PR from your main branch I'm unable to push changes here directly, if you want you can merge #4888 into this PR real quick

Everything good then Matt thank you so much, i'm ready to take another good first issue if you can open one please :)

File: crates/net/discv4/src/lib.rs [hunk 3]
```diff
@@ -1644,6 +1662,8 @@ pub enum Discv4Event {
     EnrRequest,
     /// A `EnrResponse` message was handled.
     EnrResponse,
+    /// A
```
Comment (mattsse) [inline crates/net/discv4/src/lib.rs:1665]:
```suggestion
    /// Service is being terminated
```

**Comment by DoTheBestToGetTheBest** [comment]
I take another good first issue if you open one today :)

**Pull request merged**